### PR TITLE
LocalizationMapper

### PR DESCRIPTION
**Pull request by IhateTrains**
No tests, but this is also the case in C++ one.

### CODE REVIEW

File: ImperatorToCK3/Mappers/Localizaton/LocalizationMapper.cs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+                    if (language == "french")
+                        newBlock.french = value;
+                    if (language == "german")
+                        newBlock.german = value;
+                    if (language == "russian")
+                        newBlock.russian = value;
+                    if (language == "simp_chinese")
+                        newBlock.simp_chinese = value;
+                    if (language == "spanish")
+                        newBlock.spanish = value;
```
Comment (Idhrendur) [inline ImperatorToCK3/Mappers/Localizaton/LocalizationMapper.cs:126]: This is pretty much exactly like the code in the previous block. Can you put it in a function that both call?

File: ImperatorToCK3/Mappers/Localizaton/LocalizationMapper.cs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+                }
+                if (string.IsNullOrEmpty(newBlock.russian)) {
+                    newBlock.russian = newBlock.english;
+                }
+                if (string.IsNullOrEmpty(newBlock.simp_chinese)) {
+                    newBlock.simp_chinese = newBlock.english;
+                }
+                if (string.IsNullOrEmpty(newBlock.spanish)) {
+                    newBlock.spanish = newBlock.english;
+                }
```
Comment (Idhrendur) [inline ImperatorToCK3/Mappers/Localizaton/LocalizationMapper.cs:159]: These are all very similar, is it possible to use a single function you call for each language?

**Review by Idhrendur — COMMENTED** [review]
I think this really needs some tests to make sure all the edge cases are handled. Also to demonstrate how ```ModifyForEveryLanguage``` is used, that one's intent is pretty unclear to me.

File: ImperatorToCK3/Mappers/Localizaton/LocalizationMapper.cs (hunk 2 above)
Comment (IhateTrains) [reply ImperatorToCK3/Mappers/Localizaton/LocalizationMapper.cs]: Done, see LocBlock.FillMissingLocsWithEnglish and LocBlock.FillMissingLocWithEnglish.

File: ImperatorToCK3/Mappers/Localizaton/LocalizationMapper.cs (hunk 1 above)
Comment (IhateTrains) [reply ImperatorToCK3/Mappers/Localizaton/LocalizationMapper.cs]: Done, see LocBlock.SetLocForLanguage.

**Comment by Idhrendur** [comment]
Ah, I see the mention of no tests. However, this is an opportunity to do better. Please take it.

**Comment by IhateTrains** [comment]
I've added the basic tests from CK3ToEU4, now I just need to add some info and tests for ModifyForEveryLanguage.

**Comment by IhateTrains** [comment]
@Idhrendur please check if ccf8552 clarifies the function's purpose.

**Comment by IhateTrains** [comment]
More tests have been added.

File: ImperatorToCK3/Mappers/Localizaton/LocalizationMapper.cs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        // });
+        public void ModifyForEveryLanguage(LocBlock otherLocBlock, LocDelegate modifyingMethod) {
+            modifyingMethod(ref english, otherLocBlock.english);
+            modifyingMethod(ref french, otherLocBlock.french);
+            modifyingMethod(ref german, otherLocBlock.german);
+            modifyingMethod(ref russian, otherLocBlock.russian);
+            modifyingMethod(ref simp_chinese, otherLocBlock.simp_chinese);
+            modifyingMethod(ref spanish, otherLocBlock.spanish);
+        }
+        public void SetLocForLanguage(string languageName, string value) {
```
Comment (Idhrendur) [inline ImperatorToCK3/Mappers/Localizaton/LocalizationMapper.cs:40]: If this is public, there should be tests for it too.

File: ImperatorToCK3/Mappers/Localizaton/LocalizationMapper.cs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+                    spanish = value;
+                    break;
+            }
+        }
+        private void FillMissingLocWithEnglish(ref string language) {
+            if (string.IsNullOrEmpty(language)) {
+                language = english;
+            }
+        }
+        public void FillMissingLocsWithEnglish() {
```
Comment (Idhrendur) [inline ImperatorToCK3/Mappers/Localizaton/LocalizationMapper.cs:67]: Ditto

File: ImperatorToCK3/Mappers/Localizaton/LocalizationMapper.cs [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+            if (!gotValue) {
+                return null;
+            }
+
+            if (!string.IsNullOrEmpty(locBlock.english) &&
+                (string.IsNullOrEmpty(locBlock.french) ||
+                string.IsNullOrEmpty(locBlock.german) ||
+                string.IsNullOrEmpty(locBlock.russian) ||
+                string.IsNullOrEmpty(locBlock.simp_chinese) ||
+                string.IsNullOrEmpty(locBlock.spanish))
```
Comment (Idhrendur) [inline ImperatorToCK3/Mappers/Localizaton/LocalizationMapper.cs:173]: It'd be worth creating a small function that returns a bool and contains all this logic.

**Review by Idhrendur — COMMENTED** [review]
Some late comments in case you ever feel like coming back for some cleanup

**Comment by IhateTrains** [comment]
I definitely will come back to LocalizationMapper when the converter is fully rewritten in C#, as I want to change how the localization's language is determined (language is defined in the yaml files, so this should be used instead of basing the language on file path).
